visualizations: fix x-axis for year in > GMT TZ

perrinjerome opened the pull request:
when timeserie has a `year` granularity, the time column is a int.
parseTimestamp understand this, but needs to be passed the unit to
return the timestamp in client timezone.

Fixes #7910 . See especially https://github.com/metabase/metabase/issues/7910#issuecomment-469501797 for more details about the problem.

perrinjerome committed:
visualizations: fix x-axis for year in > GMT TZ

when timeserie has a `year` granularity, the time column is a int.
parseTimestamp understand this, but needs to be passed the unit to
return the timestamp in client timezone.

perrinjerome commented:
( I removed the PR template from the description )

/cc @salsakran

perrinjerome commented:
Apparently, this cause`(0 , i.default)(...).year(...).startOf is not a function` when date is null, I need to check more.

camsaul reviewed (changes requested):
- [ ] Resolve merge conflict(s)
- [ ] Resolve "this causes `(0 , i.default)(...).year(...).startOf is not a function` when date is null."

perrinjerome commented:
I'll try to revisit this in the next few days, if I can find the time.

For the records, I was not confortable with the fact that this is a fix for a problem only happening when the browser uses some specifc timezones (the ones after GMT) and that I could not find a way to add a non-regression test reproducing the issue. In other words, I was waiting that https://github.com/metabase/metabase/pull/7247 , because it seems that with these changes there would be a way to run the frontend tests under a different timezone, which would make it possible to reproduce issue.

In any case, I will try first to rebase the changes and see if this MR still makes sense. Thanks  !

mazameli commented:
Thanks, @perrinjerome!

perrinjerome commented:
>  Resolve merge conflict(s)

I rebased these changes on current master and push forced in this branch, after confirming that #7910 still happens on master.

>  Resolve "this causes (0 , i.default)(...).year(...).startOf is not a function when date is null."

This was because when a row is null `parseTimeStamp` was called with `parseTimeStamp(null, "year")` which called:

https://github.com/metabase/metabase/blob/392ceddf640db338721d9d89082c275b9d6a9349/frontend/src/metabase/lib/time.js#L51

https://github.com/metabase/metabase/blob/392ceddf640db338721d9d89082c275b9d6a9349/frontend/src/metabase/lib/time.js#L5-L8

what happens is something like this:
```
> moment().year(null)
< 2019
> moment().year(null).startOf('year')
< Uncaught TypeError: moment(...).year(...).startOf is not a function
```

to prevent this, I updated the patch to pass the first non-null row to `parseTimeStamp`.

paulrosenzweig commented:
Closing this since the bug was fixed in #11111. We no longer use `dataOffset` and instead explicitly pass down timezone data from the server.
